utils/src/shutdown.rs

```diff
@@ -0,0 +1,163 @@
+        *self.inner.trigger.borrow()
+    }
+
+    /// Send the shutdown signal
+    #[inline]
+    pub fn shutdown(&self) {
+        if self.inner.trigger.send(Signal::SelfShutdown).is_err() {
+            warn!("no listener waiting for shutdown");
+        };
+    }
+
+    /// Check if the mpsc channel and sync follower daemon has been shutdown.
+    #[inline]
+    #[must_use]
+    pub fn check_partial_shutdown(&self) -> bool {
+        self.inner.mpmc_channel_shutdown.load(Ordering::Relaxed)
+            && self
+                .inner
+                .sync_follower_daemon_shutdown
+                .load(Ordering::Relaxed)
+    }
+
+    /// Mark mpsc channel shutdown.
+    #[inline]
+    pub fn mark_channel_shutdown(&self) {
+        self.inner
+            .mpmc_channel_shutdown
+            .store(true, Ordering::Relaxed);
+        if self.check_partial_shutdown() {
+            self.shutdown();
+        }
+    }
+
+    /// Mark sync daemon shutdown.
+    #[inline]
+    pub fn mark_sync_daemon_shutdown(&self) {
+        self.inner
+            .sync_follower_daemon_shutdown
+            .store(true, Ordering::Relaxed);
+        if self.check_partial_shutdown() {
+            self.shutdown();
+        }
+    }
+
+    /// Send the shutdown signal
+    #[inline]
+    pub fn cluster_shutdown(&self) {
+        if self.inner.trigger.send(Signal::ClusterShutdown).is_err() {
+            warn!("no listener waiting for shutdown");
+        };
+    }
+
+    /// Send the shutdown signal and wait for all listeners drop.
+    #[inline]
+    pub async fn shutdown_and_wait(&self) {
+        if self.inner.trigger.send(Signal::SelfShutdown).is_err() {
+            warn!("no listener waiting for shutdown");
+            return;
+        };
+        self.inner.trigger.closed().await;
+    }
+
+    /// Creates a new `Listener` for this `Trigger`.
+    #[inline]
+    #[must_use]
+    pub fn subscribe(&self) -> Listener {
+        Listener {
+            listener: self.inner.trigger.subscribe(),
+        }
+    }
+}
+
+/// Shutdown listener.
+#[derive(Debug, Clone)]
+pub struct Listener {
+    /// The watch receiver.
+    listener: watch::Receiver<Signal>,
+}
+
+impl Listener {
+    /// Wait for the shutdown signal.
+    #[inline]
+    pub async fn wait(&mut self) -> Option<Signal> {
+        self.listener.changed().await.ok()?;
+        Some(*self.listener.borrow())
+    }
+
+    /// Wait for the shutdown signal.
+    #[inline]
+    pub async fn wait_self_shutdown(&mut self) {
+        loop {
+            let _ig = self.listener.changed().await;
+            if matches!(*self.listener.borrow(), Signal::SelfShutdown) {
+                break;
+            }
+        }
+    }
+
+    /// Check if the shutdown signal has been sent.
+    #[inline]
+    pub fn is_shutdown(&mut self) -> bool {
+        !matches!(*self.listener.borrow(), Signal::Running)
+    }
+}
+
+/// Create a channel for shutdown.
+#[must_use]
+#[inline]
+pub fn channel() -> (Trigger, Listener) {
```

Comment:
We can acquire Listeners through `Trigger::subscribe`, so could we only return the Trigger here?

Reply:
The internal watch channel is created in pairs, and I think it is better to give the upper layer the choice than to discard it directly.